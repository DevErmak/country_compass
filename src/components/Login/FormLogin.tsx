import './form-login.css';
import { SubmitHandler, useForm } from 'react-hook-form';
import { ZodError, isValid, z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { addFavoriteCountry, setAuthentication, setUserName } from '../../store/user/infoUserSlice';
import { useDispatch } from 'react-redux';
import { useEffect, useState } from 'react';
type Props = {};

export default function FormLogin({}: Props) {
  const [isActiveAnimate, setActiveAnimate] = useState(true);
  const dispatch = useDispatch();

  const formSchemaLogin = z.object({
    login: z.string().trim().min(1, { message: 'Field is required' }).email('Email is not correct'),
    password: z.string().min(1, { message: 'Field is required' }),
  });

  type LoginFields = z.infer<typeof formSchemaLogin>;

  const {
    register,
    handleSubmit,
    reset,
    setFocus,
    trigger,
    formState: { isDirty, isSubmitting, errors, isValid },
  } = useForm<LoginFields>({
    mode: 'all',
    resolver: zodResolver(formSchemaLogin),
  });

  const onSubmit: SubmitHandler<LoginFields> = (data) => {
    let userInfo = JSON.parse(localStorage.getItem('userInfo') as string);
    if (userInfo !== null) {
      const user = userInfo.find((item: any) => item.name === data.login);
      if (user === undefined) {
        setActiveAnimate(false);
        setTimeout(() => setActiveAnimate(true), 600);
      } else {
        if (user.password === data.password) {
          dispatch(setUserName(user.name));
          console.log('32!--------------->user.listFavorite', user.listFavorite);
          dispatch(addFavoriteCountry(user.listFavorite));
          dispatch(setAuthentication(true));
        } else {
          setActiveAnimate(false);
          setTimeout(() => setActiveAnimate(true), 600);
        }
      }
    }
    reset();
  };

  useEffect(() => {
    setFocus('login');
  }, []);

  return (
    <div className="container-sign-in">
      <div className="sign-in">Sign in</div>
      <form onSubmit={handleSubmit(onSubmit)} className="form-sign-in">
        <div>
          <div className={'error-login'}>{errors.login && errors.login?.message}</div>
          <label
            htmlFor="input-login"
            className={errors.login ? 'label-login not-valid' : 'label-login'}
          >
            Login:
          </label>
          <input
            {...register('login')}
            type="text"
            className={
              isActiveAnimate ? 'input-login-sign-in' : 'input-login-sign-in input-not-valid'
            }
          />
        </div>
        <div>
          <div className={'error-password'}>{errors.password && errors.password?.message}</div>
          <label
            htmlFor="password"
            className={errors.password ? 'label-password not-valid' : 'label-password'}
          >
            Password:
          </label>
          <input
            {...register('password')}
            type="password"
            className={
              isActiveAnimate ? 'input-password-sign-in' : 'input-password-sign-in input-not-valid'
            }
          />
        </div>
        <div className="btn-sign-in">
          <button type="submit" id="btn-sign-in" disabled={!isValid}>
            Sign in
          </button>
        </div>
      </form>
    </div>
  );
}
