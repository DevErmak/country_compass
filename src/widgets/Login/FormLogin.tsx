import './form-login.scss';
import { SubmitHandler, useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { useEffect, useState } from 'react';
import { LOGIN } from '../../shared/api/graphqlV1';
import { useLazyQuery } from '@apollo/client';
import { useCookies } from 'react-cookie';
import { LoginFields, formSchemaLogin } from './model/loginFormSchema';
import Loader from '../../shared/ui/Loader';
import { setAuthentication } from '../../entities/viewer/model/user/infoUserSlice';
import { useDispatch } from 'react-redux';
import { log } from 'console';
type Props = {};

const FormLogin: React.FC<any> = ({}: Props) => {
  const [isActiveAnimate, setActiveAnimate] = useState(false);
  useEffect(() => {
    setFocus('login');
  }, []);

  const [cookie, setCookie] = useCookies(['accessToken']);

  const [loginUser, { loading, error, data: dataToken }] = useLazyQuery(LOGIN);
  // useEffect(() => {
  //   setActiveAnimate(false);
  //   setTimeout(() => setActiveAnimate(true), 600);
  // }, [error]);
  // useEffect(() => {
  //   console.log('!!!!---------------->dataToken', dataToken);
  //   if (dataToken !== undefined) {
  //     if (dataToken.login !== null) {
  //       if (dataToken.login.__typename === 'AccessToken') {
  //         console.log('kyky---------------->', dataToken.login.token);
  //         setCookie('accessToken', dataToken.login.token);
  //       }
  //     }
  //   }
  // }, [dataToken?.login]);

  const {
    register,
    handleSubmit,
    reset,
    setFocus,
    trigger,
    formState: { isDirty, isSubmitting, errors, isValid },
  } = useForm<LoginFields>({
    mode: 'all',
    resolver: zodResolver(formSchemaLogin),
  });

  const onSubmit: SubmitHandler<LoginFields> = async (data) => {
    console.log('---------------->qqqqqq');
    // const { data: dataToken, error:   } =
    await loginUser({
      variables: {
        login: data.login,
        password: data.password,
      },
    });
    if (error) {
      setActiveAnimate(false);
      setTimeout(() => setActiveAnimate(true), 600);
    }
    console.log('---------------->data', dataToken);
    if (data) setCookie('accessToken', dataToken.login.token);

    reset();
  };
  if (loading) return <Loader />;
  else
    return (
      <div className="container-sign-in">
        <div className="sign-in">Sign in</div>
        <form onSubmit={handleSubmit(onSubmit)} className="form-sign-in">
          <div>
            <div className={'error-login'}>{errors.login && errors.login?.message}</div>
            <label
              htmlFor="input-login"
              className={errors.login ? 'label-login not-valid' : 'label-login'}
            >
              Login:
            </label>
            <input
              {...register('login')}
              type="text"
              className={
                isActiveAnimate ? 'input-login-sign-in' : 'input-login-sign-in input-not-valid'
              }
            />
          </div>
          <div>
            <div className={'error-password'}>{errors.password && errors.password?.message}</div>
            <label
              htmlFor="password"
              className={errors.password ? 'label-password not-valid' : 'label-password'}
            >
              Password:
            </label>
            <input
              {...register('password')}
              type="password"
              className={
                isActiveAnimate
                  ? 'input-password-sign-in'
                  : 'input-password-sign-in input-not-valid'
              }
            />
          </div>
          <div className="btn-sign-in">
            <button type="submit" id="btn-sign-in" disabled={!isValid}>
              Sign in
            </button>
          </div>
        </form>
      </div>
    );
};
export default FormLogin;
