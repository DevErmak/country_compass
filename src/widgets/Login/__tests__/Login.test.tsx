import { renderComponentWithStore } from '../../../shared/lib/test-helpers';
import FormLogin from '../FormLogin';

it('renders correctly FormLogin', () => {
  const { container } = renderComponentWithStore({
    Component: FormLogin,
    preloadedState: {
      infoCountries: {
        listCountries: [],
        fullInfoCountry: [],
        isFullInfoCountry: false,
        isLoading: false,
        infoErrorResponse: '',
      },
      infoUser: {
        userName: '',
        listFavoriteCountries: [],
        isAuthentication: false,
        infoModal: { isActiveModal: false, formModal: 'login' },
      },
    },
  });
  expect(container).toMatchSnapshot();
});
