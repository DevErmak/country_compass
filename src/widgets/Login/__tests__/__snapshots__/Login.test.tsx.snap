// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly FormLogin 1`] = `
<div>
  <div
    class="container-sign-in"
  >
    <div
      class="sign-in"
    >
      Sign in
    </div>
    <form
      class="form-sign-in"
    >
      <div>
        <div
          class="error-login"
        />
        <label
          class="label-login"
          for="input-login"
        >
          Login:
        </label>
        <input
          class="input-login-sign-in"
          name="login"
          type="text"
        />
      </div>
      <div>
        <div
          class="error-password"
        />
        <label
          class="label-password"
          for="password"
        >
          Password:
        </label>
        <input
          class="input-password-sign-in"
          name="password"
          type="password"
        />
      </div>
      <div
        class="btn-sign-in"
      >
        <button
          disabled=""
          id="btn-sign-in"
          type="submit"
        >
          Sign in
        </button>
      </div>
    </form>
  </div>
  <div
    id="portal"
  />
</div>
`;
